app/providers/atmosphere/optimization_strategy/default.rb

```diff
@@ -64,41 +86,67 @@ def preferences
-                  f.memory >= min_mem &&
-                  f.cpu >= min_cpu &&
-                  f.hdd >= min_hdd
-                end
-              ) {|f| tmpl.get_hourly_cost_for(f) || Float::INFINITY }
-            ).sort!{ |x,y| y.memory <=> x.memory }.last
-            hsh[opt_fl] = tmpl if opt_fl
-            hsh
+          # We are given a list of tmpls which MAY be spawned by the user in the context of @appliance.
+          # We are asked to return a single template, a specific flavor and a specific tenant for instantiation.
+          # Flavor prices may vary by tenant.
+          # !CAUTION! @appliance can be nil -- this will happen when the method is invoked prior to
+          # creation of an appliance.
+          best_template = nil
+          best_tenant = nil
+          best_flavor = nil
+          instantiation_cost = Float::INFINITY
+
+          tmpls.each do |tmpl|
+            # Assuming tmpl is selected, figure out which tenant and which flavor to use.
+            candidate_tenants = get_candidate_tenants_for_template(tmpl)
+            # For each candidate tenant find optimal flavor and the corresponding VM instantiation cost
+            # Note that candidate_tenants can be blank - we must handle this case.
+            opt_tenant = nil
+            opt_flavor_and_cost = [nil, Float::INFINITY]
+            candidate_tenants.each do |t|
+              opt_flavor, cost = get_optimal_flavor_for_tenant(tmpl, t)
+              if cost < opt_flavor_and_cost[1]
+                opt_tenant = t
+                opt_flavor_and_cost = [opt_flavor, cost]
+              end
+            end
+            # An optimal tenant may not have been found - in that case skip tmpl entirely.
+            # Otherwise check if the tenant that has been found is better than the current "leader"
+            if opt_tenant.present?
+              if opt_flavor_and_cost[1] < instantiation_cost
+                best_template = tmpl
+                best_tenant = opt_tenant
+                best_flavor = opt_flavor_and_cost[0]
+                instantiation_cost = opt_flavor_and_cost[1]
+              end
+            end
           end
 
-          globally_opt_flavor = (
-            min_elements_by(
-              opt_flavors_and_tmpls_map.keys
-            ) { |f| tmpls.first.get_hourly_cost_for(f)}
-          ).sort { |x,y| x.memory <=> y.memory }.last
+          # It is possible that nothing has been found - this would indicate unsatisfiable hardware requirements
+          # (i.e. no flavor with sufficient CPU/memory/HDD) or non-existence of flavors supporting the selected
+          # architecture.
+          # If this occurs, return first template to avoid breaking user notification logic
+          # in Atmosphere::Cloud::SatisfyAppliance
+          # TODO: This is quite hacky; a better way to communicate errors to the user is needed.
+          if best_template.blank?
+            best_template = tmpls.first
+          end
 
           [
```

Comment:
I know that it is not in the scope of this PR but maybe you remember why we are returning array here instead of single the best value?